src/App.tsx

```diff
@@ -1,26 +1,201 @@
+import { TodoHeader } from './TodoHeader/TodoHeader';
+import { TodoErrors } from './TodoErrors/TodoErrors';
+import { getFilteredTodos } from './components/filteredTodos';
 
 export const App: React.FC = () => {
+  const [todos, setTodos] = useState<Todo[]>([]);
+  const [newTodosTitle, setNewTodosTitle] = useState<string>('');
+  const [errorMessage, setErrorMessage] = useState('');
+  const [filter, setFilter] = useState<FilterTypes>(FilterTypes.All);
+  const [isSubmitting, setIsSubmitting] = useState(false);
+  const inputRef = useRef<HTMLInputElement>(null);
+  const [fakeTodo, setFakeTodo] = useState<Todo | null>(null);
+  const [isLoading, setIsLoading] = useState(false);
+  const [loadingTodoId, setLoadingTodoId] = useState<number[]>([]);
+
+  useEffect(() => {
+    getTodos()
+      .then(setTodos)
+      .catch(() => setErrorMessage('Unable to load todos'));
+  }, []);
+
+  useEffect(() => {
+    if (errorMessage) {
+      const timer = setTimeout(() => {
+        setErrorMessage('');
+      }, 3000);
+
+      return () => clearTimeout(timer);
+    }
+
+    return undefined;
+  }, [errorMessage]);
+
+  useEffect(() => {
+    if (inputRef.current) {
+      inputRef.current.focus();
+    }
+  }, [todos, errorMessage]);
+
   if (!USER_ID) {
     return <UserWarning />;
   }
 
+  function addTodo({ title, userId, completed }: Omit<Todo, 'id'>) {
+    const trimmedTitle = title.trim();
+
+    if (trimmedTitle === '') {
+      setErrorMessage('Title should not be empty');
+      setIsSubmitting(false);
+
+      return;
+    }
+
+    const tempTodo: Todo = {
+      id: 0,
+      title,
+      userId,
+      completed,
+    };
+
+    setFakeTodo(tempTodo);
+    setLoadingTodoId(prevIds => [...prevIds, tempTodo.id]);
+    setIsLoading(true);
+    createTodos({ title: trimmedTitle, userId, completed })
+      .then(newTodo => {
+        setFakeTodo(null);
+        setTodos(currentTodos => [
+          ...currentTodos,
+          {
+            id: newTodo.id,
+            title: trimmedTitle,
+            userId: newTodo.userId,
+            completed: newTodo.completed,
+          },
+        ]);
+        setNewTodosTitle('');
+        setIsSubmitting(false);
+      })
+      .catch(() => {
+        setErrorMessage('Unable to add a todo');
+        setLoadingTodoId(prevIds => prevIds.filter(id => id !== tempTodo.id));
+        setFakeTodo(null);
+        setIsLoading(false);
+        setIsSubmitting(false);
+      });
+  }
+
+  function deleteTodo(todoId: number) {
+    setLoadingTodoId(prevIds => [...prevIds, todoId]);
+    deleteTodos(todoId)
+      .then(() => {
+        setTodos(currentTodos =>
+          currentTodos.filter(todo => todo.id !== todoId),
+        );
+      })
+      .catch(() => {
+        setErrorMessage('Unable to delete a todo');
+      })
+      .finally(() => {
+        setLoadingTodoId(prevIds => prevIds.filter(id => id !== todoId));
+      });
+  }
+
+  function toggleTodoCompleted(todoId: number) {
```

Comment:
It's important to stick to the same way of creating functions everywhere, fix it everywhere
```suggestion
  const toggleTodoCompleted = (todoId: number) => {
```